### ISSUE TITLE
Online Topic Modeling

### ISSUE THREAD

**Issue by AIRobotZhang** [MaartenGr/BERTopic#832]
When I use Online Topic Modeling function, I found ``Cluster reduced embeddings" step becomes slower gradually.
Is there any optimization for this problem?

**Comment by MaartenGr** [comment]
Could you share how you came to this conclusion and the corresponding code? It would greatly help in identifying what exactly is going on here.

**Comment by AIRobotZhang** [comment]
> how you came to this conclusion and the corresponding code? It would greatly help in identifying what exactl

We enter the _bertopic.py and computing the time of each step, the results are as follows:

![image](https://user-images.githubusercontent.com/20748608/201036821-2be09a96-ba6a-4613-b27f-1e83e45a2eb8.png)

![image](https://user-images.githubusercontent.com/20748608/201036919-4e8ad65b-6932-4ae3-b64e-9b4b918cedc8.png)

Extract Embeddings: 4.638892412185669
Reduce dimensionality: 12.81069803237915
Cluster reduced embeddings: 0.48001956939697266
Map and find new topics: 0.0007848739624023438
all: 18.073947429656982
#########
Extract Embeddings: 0.4368565082550049
Reduce dimensionality: 15.163034915924072
Cluster reduced embeddings: 0.727790355682373
Map and find new topics: 0.0006923675537109375
all: 16.513023614883423
#########
Extract Embeddings: 0.4288187026977539
Reduce dimensionality: 7.92698335647583
Cluster reduced embeddings: 0.9789018630981445
Map and find new topics: 0.0007781982421875
all: 9.548270463943481
#########
Extract Embeddings: 0.6770343780517578
Reduce dimensionality: 8.593830108642578
Cluster reduced embeddings: 2.2600502967834473
Map and find new topics: 0.0008578300476074219
all: 11.880823612213135
#########
Extract Embeddings: 0.6642580032348633
Reduce dimensionality: 7.263609170913696
Cluster reduced embeddings: 3.4641027450561523
Map and find new topics: 0.0010654926300048828
all: 11.803527355194092
#########
Extract Embeddings: 0.6200122833251953
Reduce dimensionality: 7.828718662261963
Cluster reduced embeddings: 4.85184383392334
Map and find new topics: 0.0010905265808105469
all: 13.854772329330444
#########
Extract Embeddings: 0.6504309177398682
Reduce dimensionality: 7.31715989112854
Cluster reduced embeddings: 5.542179346084595
Map and find new topics: 0.0011975765228271484
all: 14.380743503570557
#########
Extract Embeddings: 0.620830774307251
Reduce dimensionality: 7.320296764373779
Cluster reduced embeddings: 7.042464017868042
Map and find new topics: 0.001165151596069336
all: 15.659186363220215
#########
Extract Embeddings: 0.6284573078155518
Reduce dimensionality: 7.31078577041626
Cluster reduced embeddings: 8.938343286514282
Map and find new topics: 0.0010075569152832031
all: 17.61799192428589
#########
Extract Embeddings: 0.6075224876403809
Reduce dimensionality: 8.580629110336304
Cluster reduced embeddings: 9.467764139175415
Map and find new topics: 0.00107574462890625
all: 19.40265440940857
#########
Extract Embeddings: 0.6250205039978027
Reduce dimensionality: 8.266821384429932
Cluster reduced embeddings: 13.038195371627808
Map and find new topics: 0.0010423660278320312
all: 22.741851091384888
#########
Extract Embeddings: 0.6071341037750244
Reduce dimensionality: 8.509192705154419
Cluster reduced embeddings: 15.254991292953491
Map and find new topics: 0.001421213150024414
all: 25.29471230506897
#########
Extract Embeddings: 0.6020240783691406
Reduce dimensionality: 7.972496509552002
Cluster reduced embeddings: 18.756189107894897
Map and find new topics: 0.0013926029205322266
all: 28.313177585601807
#########
Extract Embeddings: 0.6922733783721924
Reduce dimensionality: 6.6824774742126465
Cluster reduced embeddings: 22.21987295150757
Map and find new topics: 0.0014731884002685547
all: 30.672512531280518
#########
Extract Embeddings: 0.6058018207550049
Reduce dimensionality: 7.290582180023193
Cluster reduced embeddings: 23.608723640441895
Map and find new topics: 0.0010766983032226562
all: 32.61962151527405
#########
Extract Embeddings: 0.6135420799255371
Reduce dimensionality: 9.417031526565552
Cluster reduced embeddings: 25.670673370361328
Map and find new topics: 0.0015146732330322266
all: 36.83913207054138
#########
Extract Embeddings: 0.6489794254302979
Reduce dimensionality: 7.156269073486328
Cluster reduced embeddings: 31.933584451675415
Map and find new topics: 0.0012450218200683594
all: 41.10071849822998


Below is our code:
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
from sklearn.datasets import fetch_20newsgroups
from sklearn.cluster import MiniBatchKMeans
from sklearn.decomposition import IncrementalPCA
from bertopic.vectorizers import OnlineCountVectorizer
from bertopic import BERTopic
from river import stream
from river import cluster
from bertopic.vectorizers import ClassTfidfTransformer
from tqdm import tqdm
import json
import time

class River:
    def __init__(self, model):
        self.model = model

    def partial_fit(self, umap_embeddings):
        for umap_embedding, _ in stream.iter_array(umap_embeddings):
            self.model = self.model.learn_one(umap_embedding)

        labels = []
        for umap_embedding, _ in stream.iter_array(umap_embeddings):
            label = self.model.predict_one(umap_embedding)
            labels.append(label)

        self.labels_ = labels
        return self

# Prepare documents
all_docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))["data"]
doc_chunks = [all_docs[i:i+1000] for i in range(0, len(all_docs), 1000)]

# Using DBSTREAM to detect new topics as they come in
cluster_model = River(cluster.DBSTREAM())
vectorizer_model = OnlineCountVectorizer()
ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True, bm25_weighting=True)

# Prepare model
topic_model = BERTopic(
    hdbscan_model=cluster_model, 
    vectorizer_model=vectorizer_model, 
    ctfidf_model=ctfidf_model,
    embedding_model="distiluse-base-multilingual-cased-v1"
)

topics = []
# Incrementally fit the topic model by training on 1000 documents at a time
for docs in tqdm(doc_chunks):
    topic_model.partial_fit(docs)
    print("#########")
    topics.extend(topic_model.topics_)

print(len(topics))
print(len(set(topics)))
topic_model.save("topic_model")

**Comment by MaartenGr** [comment]
It might be as a result of the `River` model that you are using. Could you also try it out with a non-River package, such as sklearn's Minibatch k-Means?

**Comment by AIRobotZhang** [comment]
> 

Thank you! But we would like to cluster without any predefined clustering number K.

**Comment by AIRobotZhang** [comment]
> It might be as a result of the `River` model that you are using. Could you also try it out with a non-River package, such as sklearn's Minibatch k-Means?

Thank you! But we would like to cluster without any predefined clustering number K.

**Comment by MaartenGr** [comment]
> Thank you! But we would like to cluster without any predefined clustering number K.

You could do a preliminary pass with a subset of the data to get a feeling of a good number of clusters using HDBSCAN and then pass Minibatch k-Means with that number of clusters.

**Comment by MaartenGr** [comment]
Due to inactivity, I'll be closing this issue for now. If, however, you want to continue the discussion or re-open the issue, feel free to reach out!
